### PR TITLE
refactor VideoViewer, AudioViewer and PhotoViewer activities 

### PR DESCRIPTION
**Pull request by wafalakhal**
Add sharedViewmodel to encapsulate common functions 
Add a common VaultActionHelper 
refactor old code  of the three activities
change the permissions requests using contracts instead of activityResults method

### CODE REVIEW

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/AudioPlayActivity.kt [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+import rs.readahead.washington.mobile.util.ThreadUtil
+import rs.readahead.washington.mobile.views.activity.MetadataViewerActivity
+import rs.readahead.washington.mobile.views.activity.viewer.VaultActionsHelper.initContracts
+import rs.readahead.washington.mobile.views.activity.viewer.VaultActionsHelper.showVaultActionsDialog
+import rs.readahead.washington.mobile.views.base_ui.BaseLockActivity
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+class AudioPlayActivity : BaseLockActivity(), StyledPlayerView.ControllerVisibilityListener {
+    var mPlay: ImageButton? = null
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/AudioPlayActivity.kt:36]: Can we make elements like this not null?

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/AudioPlayActivity.kt [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+    private fun disableScreenTimeout() {
+        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
+    }
+
+    private fun enableScreenTimeout() {
+        window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
+    }
+
+    private fun initView() {
+        mPlay = binding!!.content.playAudio
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/AudioPlayActivity.kt:444]: Assertions like this are unnecessary

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/AudioPlayActivity.kt [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+    }
+
+    private fun disablePlay() {
+        mPlay!!.setImageDrawable(this.resources.getDrawable(R.drawable.big_white_pause_24p))
+        enableButton(forward, mFwd)
+        enableButton(rewind, mRwd)
+    }
+
+    private fun enablePlay() {
+        mPlay!!.setImageDrawable(this.resources.getDrawable(R.drawable.ic_play_arrow_white_24dp))
```
Comment (djordjemarusic) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/AudioPlayActivity.kt:387]: this getDrawable call is depricated

**Comment by djordjemarusic** [comment]
In general, we are also cleaning the code so warnings should be handled in the classes we are working on.

**Review by djordjemarusic — APPROVED** [review]

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/PhotoViewerActivity.kt [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+import rs.readahead.washington.mobile.bus.event.VaultFileRenameEvent
+import rs.readahead.washington.mobile.databinding.ActivityPhotoViewerBinding
+import rs.readahead.washington.mobile.media.MediaFileHandler
+import rs.readahead.washington.mobile.views.activity.MetadataViewerActivity
+import rs.readahead.washington.mobile.views.activity.viewer.VaultActionsHelper.initContracts
+import rs.readahead.washington.mobile.views.activity.viewer.VaultActionsHelper.showVaultActionsDialog
+import rs.readahead.washington.mobile.views.base_ui.BaseLockActivity
+
+
+class PhotoViewerActivity : BaseLockActivity(), StyledPlayerView.ControllerVisibilityListener {
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/PhotoViewerActivity.kt:34]: We should use hilt to inject the `ViewModel` and other dependencies.

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/PhotoViewerActivity.kt [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    private lateinit var binding: ActivityPhotoViewerBinding
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityPhotoViewerBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        initContracts()
+        setupToolbar()
+        initObservers()
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+            binding.toolbar.outlineProvider = null
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/PhotoViewerActivity.kt:56]: There should be an extension function in Djordje's PR now, that you can use configure the appBar

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/SharedMediaFileViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+            })
+    }
+
+    /**
+     * Deletes the specified [vaultFile] from the application's media storage.
+     *
+     * @param vaultFile The VaultFile to be deleted.
+     */
+    fun deleteMediaFiles(vaultFile: VaultFile?) {
+        disposables.add(MyApplication.rxVault.delete(vaultFile)
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/SharedMediaFileViewModel.kt:122]: Why `VaultFile`is nullable? if needed we should have an early return no?

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/SharedMediaFileViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        )
+    }
+
+    /**
+     * Retrieves a VaultFile based on its [id].
+     *
+     * @param id The ID of the VaultFile to retrieve.
+     */
+    fun getMediaFile(id: String?) {
+        disposables.add(
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/SharedMediaFileViewModel.kt:180]: Same here why the is nullable, let's have an early return, and apply the logic on the rest of the code please.

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/VaultActionsHelper.kt [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+    }
+
+
+    // Check if the app has storage permissions
+    private fun hasStoragePermissions(context: Context): Boolean {
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+            Environment.isExternalStorageManager()
+        } else {
+            val result: Int = ContextCompat.checkSelfPermission(
+                context, Manifest.permission.READ_EXTERNAL_STORAGE
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/VaultActionsHelper.kt:326]: Do you think we should have a separate helper to hold the permissions?

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/VideoViewerActivity.kt [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+import rs.readahead.washington.mobile.databinding.ActivityVideoViewerBinding
+import rs.readahead.washington.mobile.media.MediaFileHandler
+import rs.readahead.washington.mobile.media.exo.MediaFileDataSourceFactory
+import rs.readahead.washington.mobile.util.DialogsUtil
+import rs.readahead.washington.mobile.views.activity.MetadataViewerActivity
+import rs.readahead.washington.mobile.views.activity.viewer.VaultActionsHelper.initContracts
+import rs.readahead.washington.mobile.views.activity.viewer.VaultActionsHelper.showVaultActionsDialog
+import rs.readahead.washington.mobile.views.base_ui.BaseLockActivity
+
+class VideoViewerActivity : BaseLockActivity(), StyledPlayerView.ControllerVisibilityListener {
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/VideoViewerActivity.kt:37]: Same here let's use dagger if needed!

File: mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/VideoViewerActivity.kt [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+        super.onCreate(savedInstanceState)
+        binding = ActivityVideoViewerBinding.inflate(layoutInflater)
+        overridePendingTransition(R.anim.slide_in_start, R.anim.fade_out)
+        setContentView(binding.root)
+        actionsDisabled = intent.hasExtra(NO_ACTIONS)
+        initContracts()
+        setupToolbar()
+        shouldAutoPlay = true
+        clearResumePosition()
+        simpleExoPlayerView = findViewById(R.id.player_view)
```
Comment (ahlem-jarrar) [inline mobile/src/main/java/rs/readahead/washington/mobile/views/activity/viewer/VideoViewerActivity.kt:71]: why we don't access this with `binding`here?

**Review by ahlem-jarrar — CHANGES_REQUESTED** [review]

**Pull request merged**